ocaml/xapi/xapi_guest_agent.ml

```diff
@@ -112,10 +124,56 @@ let networks path (list: string -> string list) =
          | Some pair -> pair :: acc
       ) [] (list path)
   in
-  path
-  |> find_eths
-  |> List.map (fun (path, prefix) -> find_all_ips path prefix)
-  |> List.concat
+  let find_vifs vif_path =
+    let extract_vif acc vif_id = ((extend vif_path vif_id), vif_id) :: acc in  
+    List.fold_left extract_vif [] (list vif_path)
+  in
+  let cmp a b = 
+    try 
+      compare (int_of_string a) (int_of_string b) 
+    with Failure _ -> 
+      error "String (\"%s\" or \"%s\") can't be converted into an integer as index of IP" a b;
+      raise (Failure "Failed to compare")
+  in
+  let find_all_vif_ips vif_path vif_id = 
+    (*  vif_path: attr/vif/0 *)
+    (*  vif_id: 0 *)
+    let extract_ip_ver vif_id acc ip_ver = 
+      let ip_addr_ids = list (extend vif_path ip_ver)  in
+      let extract_ip_addr vif_id ip_ver acc ip_addr_id = 
+        let key_left = Printf.sprintf "%s/%s/%s" vif_path ip_ver ip_addr_id in
+        let key_right = Printf.sprintf "%s/%s/%s" vif_id ip_ver ip_addr_id in
+        match acc with
+        | [] when ip_ver = "ipv4"  -> 
+          [(key_left, (extend vif_id "ip")); (key_left, key_right)]
+        | _ -> (key_left, key_right) :: acc
+      in  
+      try 
+        List.fold_left (extract_ip_addr vif_id ip_ver) 
+          acc (List.stable_sort cmp ip_addr_ids)
+      with Failure _ ->
+        error "Failed to extract IP address for vif %s." vif_id;
+        []
+    in
+    let ip_vers = List.filter (fun a -> a = "ipv4" || a = "ipv6") (list vif_path) in
+    List.fold_left (extract_ip_ver vif_id) [] ip_vers
+  in
+  let not_empty = function [] -> false | _ -> true in
+  let iface_pair_list = find_eths path in
+  let vif_pair_list = find_vifs (extend path "vif") in
+  match not_empty vif_pair_list, not_empty iface_pair_list with
+  | (true, false) ->
+    vif_pair_list
+    |> List.map (fun (vif_path, vif_id) -> find_all_vif_ips vif_path vif_id)
+    |> List.concat
+  | (false, true) ->
+    iface_pair_list
+    |> List.map (fun (path, prefix) -> find_all_ips path prefix)
+    |> List.concat
+  | (true, true) ->
+    error "Both vif and iface(ethX) were found under %s in xenstore." path; []
+  | (false, false) ->
+    error "Neither vif nor iface(ethX) was found under %s in xenstore." path; []
```

Comment:
I don't think this is an error: the VM may not actually have any IP addresses (yet). We should just return `[]`.